package chimerascan_test_setup;

use strict;
use warnings;

BEGIN {
    $ENV{UR_DBI_NO_COMMIT} = 1;
    $ENV{UR_USE_DUMMY_AUTOGENERATED_IDS} = 1;
};

use above 'Genome';
require Exporter;
our @ISA = qw(Exporter);
our @EXPORT_OK = qw(setup);

sub setup {
    my $data_dir = $ENV{GENOME_TEST_INPUTS} . "/Genome-Model-RnaSeq-DetectFusionsResult-ChimerascanResult/0.4.5";
    my $tophat_dir = $data_dir . "/tophat_data4";

    my $t = Genome::Taxon->__define__(name => 'human');
    my $p = Genome::Individual->create(name => "test-human-patient", common_name => 'testpatient', taxon => $t);
    my $s = Genome::Sample->create(name => "test-human-patient", common_name => 'tumor', source => $p);
    my ($species_names, $versions) = @_;

    my $ref_pp = Genome::ProcessingProfile::ImportedReferenceSequence->create(name => 'test_ref_pp');

    my $ref_model = Genome::Model::ImportedReferenceSequence->create(
        name                => "test_ref_sequence_human",
        processing_profile  => $ref_pp,
        subject_class_name  => ref($s),
        subject_id          => $s->id,
    );

    my $reference_build = Genome::Model::Build::ImportedReferenceSequence->create(
        name            => "ref_sequence_$s-37",
        model           => $ref_model,
        fasta_file      => 'turkey_sammich',
        data_directory  => "$data_dir/ref",
        version         => '37',
    );


    my $annotation_model = Genome::Model::ImportedAnnotation->create(
        name => '1 chr test annotation',
        subject => $ref_model->subject,
        processing_profile => Genome::ProcessingProfile->get(name => 'imported-annotation.ensembl'),
        reference_sequence => $reference_build,
    );

    my $annotation_build = Genome::Model::Build::ImportedAnnotation->__define__(
        version => 'v1',
        model => $annotation_model,
        data_directory => '/gscmnt/gc8002/info/model_data/2772828715/build125092315', #65_37j_v6
    );

    my $alignment_result = Genome::InstrumentData::AlignmentResult::Tophat->__define__(
        aligner_name => 'tophat',
        output_dir => $tophat_dir,
        reference_build_id => $reference_build->id,
        bowtie_version => '0.12.7'
    );
    $alignment_result->lookup_hash($alignment_result->calculate_lookup_hash());

    my $index = Genome::Model::RnaSeq::DetectFusionsResult::ChimerascanResult::Index->__define__(
        version => "0.4.5",
        bowtie_version => "0.12.7",
        reference_build => $reference_build,
        output_dir => $data_dir.'/IndexResult/',
        annotation_build => $annotation_build,
    );
    $index->lookup_hash($index->calculate_lookup_hash());

    return $alignment_result, $annotation_build;
}

1;
